### PR TITLE
Fix S107 FP: Don't raise on Zero-overhead member access methods

### LINKED ISSUE #8156: Fix S107 FP: Don't raise on Zero-overhead member access methods
A `static extern` method used as [C# 12 Zero-overhead member access](https://github.com/dotnet/runtime/issues/81741) method needs to match the signature of its target method or constructor.

If the target method/constructor has too many parameters (or exactly the maximum number of parameters), the zero-overhead member access method needs to match that or have one more (in the case of constructors). That results in S107 not being respected.

If the target method is not under the control of the owner of the zero-overhead member access method, there is nothing he can do to fix the issue, since signatures have to match.

For this reason, there should be an exception in presence of an `UnsafeAccessorAttribute`.

```cs
using System.Runtime.CompilerServices;

class ZeroOverheadMemberAccess
{
    [UnsafeAccessor(UnsafeAccessorKind.Constructor)]
    extern static UserData CallConstructor(int x1, int x2, int x3, int x4);                 // Noncompliant, FP: signature has to match target

    [UnsafeAccessor(UnsafeAccessorKind.Method, Name = "Method")]
    extern static void CallMethod(UserData userData, int x1, int x2, int x3, int x4);       // Noncompliant, FP: signature has to match target

    [UnsafeAccessor(UnsafeAccessorKind.StaticMethod, Name = "StaticMethod")]
    extern static void CallStaticMethod(UserData userData, int x1, int x2, int x3, int x4); // Noncompliant, FP: signature has to match target
}

class UserData
{
    UserData(int x1, int x2, int x3, int x4) { }         // Noncompliant

    void Method(int x1, int x2, int x3) { }              // Compliant
    static void StaticMethod(int x1, int x2, int x3) { } // Compliant
}
```

### PR DESCRIPTION
**Pull request by cristian-ambrosini-sonarsource**
Fixes #8156

[Commit with implementation](https://github.com/SonarSource/sonar-dotnet/pull/8254/commits/56c7b0829f05fe3140ea35f46bdf715996368aeb).
[Commit with tests](https://github.com/SonarSource/sonar-dotnet/pull/8254/commits/36b455b59a50648fa4c1a66af0865664994d83cc).

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/TooManyParameters.cs [hunk 1]
```diff
@@ -38,6 +38,12 @@ public class TooManyParameters : TooManyParametersBase<SyntaxKind, ParameterList
             { SyntaxKindEx.LocalFunctionStatement, "Local function" }
         }.ToImmutableDictionary();
 
+        protected override bool IsZeroOverheadMemberAccess(SyntaxNode node) =>
+            node is MethodDeclarationSyntax { AttributeLists: { Count: > 0 } attributeList } methodDeclaration
+            && methodDeclaration.Modifiers.Any(SyntaxKind.ExternKeyword)
+            && methodDeclaration.Modifiers.Any(SyntaxKind.StaticKeyword)
+            && attributeList.Any(x => x.Attributes.Any(y => y.Name.ToString() == "UnsafeAccessor"));
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/TooManyParameters.cs:45]: Use `AttributeListSyntaxExtensions.GetAttributes`

File: analyzers/tests/SonarAnalyzer.Net8.UnitTest/TestCases/TooManyParameters.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+﻿using System.Runtime.CompilerServices;
+
+class ZeroOverheadMemberAccess
+{
+    [UnsafeAccessor(UnsafeAccessorKind.Constructor)]
+    extern static UserData CallConstructor(int x1, int x2, int x3, int x4);                 // Compliant: signature has to match target
+
+    [UnsafeAccessor(UnsafeAccessorKind.Method, Name = "Method")]
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.Net8.UnitTest/TestCases/TooManyParameters.cs:8]:
```suggestion
    [UnsafeAccessorAttribute(UnsafeAccessorKind.Method, Name = "Method")]
```

File: analyzers/tests/SonarAnalyzer.Net8.UnitTest/TestCases/TooManyParameters.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    extern static void CallMethod(UserData userData, int x1, int x2, int x3, int x4);       // Compliant: signature has to match target
+
+    [UnsafeAccessor(UnsafeAccessorKind.StaticMethod, Name = "StaticMethod")]
+    extern static void CallStaticMethod(UserData userData, int x1, int x2, int x3, int x4); // Compliant: signature has to match target
+}
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.Net8.UnitTest/TestCases/TooManyParameters.cs:13]: There is no need to special case `UnsafeAccessor`. `extern` should be enough:
```suggestion
    [DllImport("user32.dll")]
    extern static void SomeMethod(UserData userData, int x1, int x2, int x3, int x4); // Compliant: signature has to match target
}
```

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
I think the problem can be solved in a more generic way, by just looking for `extern`. Anything that is defined `extern` has to have a matching signature to something that is defined elsewhere. Therefore the developer has no chance to change it.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/TooManyParameters.cs (hunk 1 above)
Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/TooManyParameters.cs]: Not checking attributes anymore

File: analyzers/src/SonarAnalyzer.VisualBasic/Facade/VisualBasicSyntaxFacade.cs [hunk 4]
```diff
@@ -88,6 +88,9 @@ internal sealed class VisualBasicSyntaxFacade : SyntaxFacade<SyntaxKind>
 
     public override bool IsAnyKind(SyntaxTrivia trivia, params SyntaxKind[] syntaxKinds) => trivia.IsAnyKind(syntaxKinds);
 
+    public override bool IsExtern(SyntaxNode node) =>
+        node is MethodBlockSyntax { } methodBlock && methodBlock.IsDeclare();
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.VisualBasic/Facade/VisualBasicSyntaxFacade.cs:92]:
```suggestion
        node is DeclareStatementSyntax || HasDllImportAttribute(node);
```

File: analyzers/src/SonarAnalyzer.VisualBasic/Extensions/MethodBlockSyntaxExtensions.cs [hunk 5]
```diff
@@ -22,6 +22,9 @@ namespace SonarAnalyzer.Extensions;
 
 public static class MethodBlockSyntaxExtensions
 {
+    public static bool IsDeclare(this MethodBlockSyntax methodBlock) =>
+        methodBlock.SubOrFunctionStatement.Modifiers.Any(SyntaxKind.DeclareKeyword);
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.VisualBasic/Extensions/MethodBlockSyntaxExtensions.cs:26]: This looks wrong to me and isn't under test. I would expect something like this:
```suggestion
    public static bool IsDeclare(this DeclareStatementSyntax method) =>
        true;
```
or this:
```cs
    public static bool IsDeclare(this MethodBaseSyntax method) =>
        method is DeclareStatementSyntax;
```

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
`IsDeclare` extension methods should be under test.

File: analyzers/src/SonarAnalyzer.Common/Rules/TooManyParametersBase.cs [hunk 6]
```diff
@@ -50,7 +51,10 @@ public abstract class TooManyParametersBase<TSyntaxKind, TParameterListSyntax> :
                 {
                     var parametersCount = CountParameters((TParameterListSyntax)c.Node);
                     var baseCount = BaseParameterCount(c.Node.Parent);
-                    if (parametersCount - baseCount > Maximum && c.Node.Parent != null && CanBeChanged(c.Node.Parent, c.SemanticModel))
+                    if (parametersCount - baseCount > Maximum
+                        && c.Node.Parent is { } parent
+                        && !IsExtern(parent)
+                        && CanBeChanged(parent, c.SemanticModel))
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/TooManyParametersBase.cs:57]: CanBeChanged retrieves the method symbol. The method symbol also has some indications of "extern":
* [IMethodSymbol.MethodImplementationFlags](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.imethodsymbol.methodimplementationflags?view=roslyn-dotnet-4.6.0)
* [IMethodSymbol.MethodKind](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.imethodsymbol.methodkind?view=roslyn-dotnet-4.6.0) (MethodKind.DeclareMethod)

I don't know if either of these properties help to identify extern methods. What do we want to do about VB support? create an issue? If so, we could add this information there.

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM. But we need a follow-up issue for VB.

File: analyzers/src/SonarAnalyzer.Common/Rules/TooManyParametersBase.cs (hunk 6 above)
Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/TooManyParametersBase.cs:57]: If you agree, given that VB is no longer a priority, I would not even create the issue and wait for the community to raise concerns.

**Pull request merged**